### PR TITLE
feat(token registry): proto max collateral share

### PR DESCRIPTION
**Pull request by robert-zaremba**
## Description

ref: #1095

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] added `!` to the type prefix if API or client breaking change
- [x] added appropriate labels to the PR
- [x] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [x] added a changelog entry to `CHANGELOG.md`
- [x] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [x] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items._

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: proto/umee/leverage/v1/leverage.proto
Comment (toteki) [inline proto/umee/leverage/v1/leverage.proto:149]:
```suggestion
    (gogoproto.moretags) = "yaml:\"max_collateral_supply\""
```

File: proto/umee/leverage/v1/leverage.proto
Comment (toteki) [inline proto/umee/leverage/v1/leverage.proto:143]:
```suggestion
  // Maximum allowed collateral supply specifies how much of the token can
```

File: proto/umee/leverage/v1/leverage.proto
Comment (toteki) [inline proto/umee/leverage/v1/leverage.proto:144]:
```suggestion
  // be provided in total as a collateral in the system.
```

**Comment by toteki** [comment]
For clarity, what is an example of this in use?

Option 1 - If I'm understanding correctly:
```
MaxCollateralSupply(tokenA) = 20%

Each time collateral of token A is added to the system, the total USD value of all collateral in the system is calculated. If the total USD value of all token A collateral in the system would exceed 20% of the first value, `MsgAddCollateral` fails.

There might also be restrictions on `MsgRemoveCollateral` of token B, if removing token B would cause token A to exceed 20% of system collateral value.
```

Option 2 - if I'm misunderstanding the wording, this was another possibility
```
MaxCollateralSupply(tokenA) = 20%

If the total supply of token A is 1000, then only 200 can be used as collateral until more is supplied.
```

Am I correct in assuming this proposal is the first option?

File: proto/umee/leverage/v1/leverage.proto
Comment (toteki) [inline proto/umee/leverage/v1/leverage.proto:146]: In cases like this, we usually go for an `sdk.Dec` ranging from `0.0` to `1.0`.

Comment (robert-zaremba) [reply proto/umee/leverage/v1/leverage.proto:146]: I was thinking about it. Personally for small, well defined values I prefer ints:
* faster (both serialization / deserialization and computation)
* smaller in storage

**Comment by robert-zaremba** [comment]
> For clarity, what is an example of this in use?

I will add ADR update once this will get merged with a proper example. Option 1 is the valid one - see the linked issue - it has more details.

> There might also be restrictions on `MsgRemoveCollateral` of token B, if removing token B would cause token A to exceed 20% of system collateral value.

This is a good and valid point. The restriction will only hold when adding a collateral.

**Comment by toteki** [comment]
We might also solicit naming advice for this parameter in discussion tomorrow.

`MaxCollateralSupply` on its own suggests a limit to the amount of collateral itself, rather than a "ratio" or "share" of the total market.

**Comment by robert-zaremba** [comment]
> MaxCollateralSupply on its own suggests a limit to the amount of collateral itself, rather than a "ratio" or "share" of the total market.

How about `MaxCollateralSupplyRatio`? Feel free to propose other names.

**Comment by toteki** [comment]
MaxCollateralSupply (wrong meaning)
MaxCollateralMarketShare? (too long)
MaxCollateralShare?

File: proto/umee/leverage/v1/leverage.proto
Comment (robert-zaremba) [reply proto/umee/leverage/v1/leverage.proto:146]: I think we kind of decided to use ints. Let's use int here - this is a new code, so doesn't change nor interact with other things. Really, no need to duplicate the work now.

In a separate PR I will update other relevant messages to use ints

**Comment by robert-zaremba** [comment]
Renamed to `MaxCollateralShare`

File: proto/umee/leverage/v1/leverage.proto
Comment (toteki) [inline proto/umee/leverage/v1/leverage.proto:145]:
```suggestion
  // Maximum collateral share specifies how much of the
  // system's overall collateral be provided by a single token.
  // Value a percent; allowed values are in [0, 100] range.
```

File: x/leverage/types/token.go
Comment (toteki) [inline x/leverage/types/token.go:88]:
```suggestion
		return sdkerrors.ErrInvalidRequest.Wrap("Token. MaxCollateralShare must be in [0; 100] range")
```

File: x/leverage/types/token_test.go
Comment (toteki) [inline x/leverage/types/token_test.go:127]:
```suggestion
	validMaxCollateralShare := validToken()
	validMaxCollateralShare.MaxCollateralShare = 100
```

File: x/leverage/types/token_test.go
Comment (toteki) [inline x/leverage/types/token_test.go:190]:
```suggestion
		"invalid max collateral share": {
			input:     invalidMaxCollateralShare,
			expectErr: true,
		},
		"valid max collateral share": {
			input:     validMaxCollateralShare,
			expectErr: false,
		},
```

File: proto/umee/leverage/v1/leverage.proto
Comment (toteki) [inline proto/umee/leverage/v1/leverage.proto:149]:
```suggestion
    (gogoproto.moretags) = "yaml:\"max_collateral_share\""
```

**Review by toteki — APPROVED** [review]
Did a little housekeeping (see commits) - approving now.

File: x/leverage/types/token_test.go
Comment (toteki) [inline x/leverage/types/token_test.go:184]:
```suggestion
```

**Pull request merged**
